src/devices/Common/Iot/Device/Common/ValueArray.cs

```diff
@@ -91,22 +91,53 @@ public int Count
         /// <inheritdoc />
         public bool IsReadOnly => false;
 
+#if NET5_0_OR_GREATER
         /// <summary>
         /// Returns a span pointing to this instance
         /// </summary>
         /// <returns>A span representing this instance</returns>
+        /// <remarks>
+        /// Due to a shortcoming of the Garbage collector, it is the caller's responsibility
+        /// to make sure the lifetime of the Span returned from this method is not longer
+        /// than the lifetime of the array. See also
+        /// https://github.com/dotnet/iot/issues/1886
+        /// </remarks>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public unsafe Span<T> AsSpan()
+        public Span<T> AsSpan()
+        {
+            return MemoryMarshal.CreateSpan(ref _e0, MaximumSize);
+        }
+#else
+        /// <summary>
+        /// Returns a span pointing to this instance
+        /// </summary>
+        /// <returns>A span representing this instance</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+#if !BUILDING_IOT_DEVICE_BINDINGS
+        public
+#else
+        internal
+#endif
+        unsafe Span<T> AsSpan()
```

Comment:
I would personally leave out `unsafe` at all, even for the netstandard2.0 case, by using `Memory<T>`.

Reply:
That would make the API very different in use for netstandard2.0 and require `#ifs` in all places where it's used, which is ugly.

Reply:
This is why I suggested in the triage to entirely replace the `Span` with `Memory` for all the cases.
Consumers can just take the `Memory<T>` (which own the memory) from which they can obtain the Span as they always did.
IMO safety should come first, otherwise you leave to the consumers the responsibility to avoid the bug, which is not necessarily easy to find.

Reply:
`Memory<T>` is much slower (removes almost all benefit of the `AsSpan` method as memory -> span is quite costly) and is also more unsafe than `Span<T>` (it can suffer from cross-thread tearing and other issues)